src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,429 @@
+
+    private final Duration maxWaitTimePerRecordForTableSideData;
+    private final Duration frequencyToCheckForExpiredWaitTimes;
+    private final String streamBufferStoreName;
+    private final String tableStoreName;
+
+    StreamTableJoinStreamWaitsForTable(final Duration maxWaitTimePerRecordForTableData,
+                                       final Duration frequencyToCheckForExpiredWaitTimes,
+                                       final String streamBufferStoreName,
+                                       final String tableStoreName) {
+      this.maxWaitTimePerRecordForTableSideData = maxWaitTimePerRecordForTableData;
+      this.frequencyToCheckForExpiredWaitTimes = frequencyToCheckForExpiredWaitTimes;
+      this.streamBufferStoreName = streamBufferStoreName;
+      this.tableStoreName = tableStoreName;
+    }
+
+    @Override
+    public Transformer<String, Double, KeyValue<String, Pair<Double, Long>>> get() {
+      return new Transformer<String, Double, KeyValue<String, Pair<Double, Long>>>() {
+
+        private KeyValueStore<String, Pair<Double, Instant>> streamBufferStore;
+        private KeyValueStore<String, Long> tableStore;
+        private ProcessorContext context;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+          tableStore = (KeyValueStore<String, Long>) context.getStateStore(tableStoreName);
+          this.context = context;
+          // Note: In practice, you will probably want to use `PunctuationType.STREAM_TIME`.  However, using stream time
+          // would make the test/validation setup in this example more complicated, hence we use wall clock time.
+          this.context.schedule(frequencyToCheckForExpiredWaitTimes, PunctuationType.WALL_CLOCK_TIME, this::punctuate);
+        }
+
+        @Override
+        public KeyValue<String, Pair<Double, Long>> transform(final String key, final Double value) {
+          return sendFullJoinRecordOrWaitForTableSide(key, value);
+        }
+
+        private KeyValue<String, Pair<Double, Long>> sendFullJoinRecordOrWaitForTableSide(final String key, final Double value) {
+          final Long tableValue = tableStore.get(key);
+          if (tableValue != null) {
+            // We have data for both the stream and the table, so we can send a fully populated join message downstream
+            // immediately.
+            return KeyValue.pair(key, new Pair<>(value, tableValue));
+          } else {
+            // Don't send a join output just yet because we're still lacking table-side information.  Instead, buffer
+            // the current stream-side record, hoping that the table side will eventually see a matching record within
+            // `maxWaitTimePerRecordForTableSideData`.
+            streamBufferStore.put(key, new Pair<>(value, Instant.now()));
```

Comment:
This won't work if a second record with the same key arrives, because it would overwrite the already buffered record resulting in data loss.

Should we put `context.timestamp()` into the store to get event-time semantics?

Reply:
Ack. I moved the example from wall clock time to event-time, but this required a few additional changes to enhance IntegrationTestUtils to explicitly control timestamps of test data.